glib-macros/src/properties.rs

```diff
@@ -0,0 +1,527 @@
+    quote!(#(#defs;)*)
+}
+
+fn expand_getset_properties_impl(imp_type_ident: &syn::Ident, props: &[PropDesc]) -> TokenStream2 {
+    let defs = props.iter().map(|p| {
+        let ident = name_to_ident(&p.name);
+        let field_ident = &p.field_ident;
+        let ty = &p.ty;
+
+        let getter = p.get.as_ref().map(|mfn| {
+            let body = match (p.member.as_ref(), mfn) {
+                (None, MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.get(|x| x.to_owned())
+                ),
+                (Some(member), MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.get(|x| x.#member.to_owned())
+                ),
+                (_, MaybeCustomFn::Custom(custom_fn)) => {
+                    let custom_fn = change_self_type(custom_fn, &imp_type_ident.to_string());
+                    quote!((#custom_fn)(&self.imp()))
+                }
+            };
+            let fn_prototype = getter_prototype(&ident, ty);
+            quote!(#fn_prototype { #body })
+        });
+        let setter = p.set.as_ref().map(|mfn| {
+            let body = match (p.member.as_ref(), mfn) {
+                (None, MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.set(move |x| *x = value)
+                ),
+                (Some(member), MaybeCustomFn::Default) => quote!(
+                     self.imp().#field_ident.set(move |x| x.#member = value)
+                ),
+                (_, MaybeCustomFn::Custom(custom_fn)) => {
+                    let custom_fn = change_self_type(custom_fn, &imp_type_ident.to_string());
+                    quote!((#custom_fn)(&self.imp(), value))
+                }
+            };
+            let fn_prototype = setter_prototype(&ident, ty);
+            quote!(#fn_prototype { #body })
+        });
+        quote!(
+            #getter
+            #setter
+        )
+    });
+    quote!(#(#defs)*)
+}
+
+fn expand_connect_prop_notify(p: &PropDesc) -> TokenStream2 {
+    let name = &p.name;
+    let fn_ident = format_ident!("connect_{}_notify", name_to_ident(name));
+    quote!(fn #fn_ident<F: Fn(&Self) + 'static>(&self, f: F) -> glib::SignalHandlerId)
+}
+fn expand_connect_prop_notify_def(props: &[PropDesc]) -> TokenStream2 {
+    let connection_fns = props.iter().map(expand_connect_prop_notify);
+    quote!(#(#connection_fns;)*)
+}
+fn expand_connect_prop_notify_impl(props: &[PropDesc]) -> TokenStream2 {
+    let connection_fns = props.iter().map(|p| {
+        let name = &p.name;
+        let fn_prototype = expand_connect_prop_notify(p);
+        quote!(#fn_prototype {
+            self.connect_notify_local(Some(#name), move |this, _| {
+                f(this)
+            })
+        })
+    });
+    quote!(#(#connection_fns)*)
+}
+
+pub fn impl_derive_props(input: PropsMacroInput) -> TokenStream {
+    let struct_ident = &input.ident;
+    let struct_ident_ext = format_ident!("{}Ext", &input.ident);
```

Comment:
For a signals macro that is separate from this, we are going to have to generate another `Ext` trait for all the `emit`s and `connect`s. So maybe this should be `PropsExt`, or have some way to customize the name of the trait?